[Woo POS] Show Location Permissions notice and error in Point of Sale

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15223 <!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Location permission is required to take card payments. Previously, the POS skipped the location pre-alert and the error for not having location permission when connecting the reader, which led to a failure when taking a payment.

This PR adds the views needed to show the location messages in POS.

Prior to this, M2 readers were unaffected, but WisePad3 readers could get stuck on an `Authorizing` step for a minute or two. That's no longer so likely, as it would require turning off location permissions after you've connected the reader.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Deny permission

1. Set the location permissions for Woo to `Ask next time` in iOS settings
2. Using a UK store, launch the app and navigate to POS
3. Tap `Connect your reader`
4. Observe that you're shown a pre-alert for the location permissions request
5. Don't allow the permissions
6. Observe that you're shown the location required error
7. Tap `x`
8. Observe that the reader does not connect.

### Grant permission

1. Set the location permissions for Woo to `Ask next time` in iOS settings
2. Using a UK store, launch the app and navigate to POS
3. Tap `Connect your reader`
4. Observe that you're shown a pre-alert for the location permissions request
5. Grant location permission
9. Add items to the cart and check out
10. Tap your card
11. Observe that payment completes successfully, and there's no payment loop

### Grant permission after a previous denial 

1. Set the location permissions for Woo to `Never` in iOS settings
2. Using a UK store, launch the app and navigate to POS
3. Tap `Connect your reader`
4. Observe that you're shown the location required error
5. Tap the `Go to settings` button
6. Observe that you're taken to the Woo screen in iOS settings
7. Grant location permission in settings
8. Go back to the app – observe that connection continues and completes
9. Add items to the cart and check out
10. Tap your card
11. Observe that payment completes successfully, and there's no payment loop

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I've tested the above flows with an iPad Air running iOS 17.7.3, and a simulator and simulated reader. With the physical readers, I tested with an M2 and a Wisepad3

There was one very tricky bug around retrying payments after the first retry attempt – the payment cancellable was populated but invalid, when it should have been nil.

This happened because the `collectPaymentMethod` call on the first retry called its completion handler (which should have nilled the cancellable) before it returned the cancellable – so the nil happened, but then the function returned and we set the cancellable back up again. Checking some cancellation flows is worthwhile, particularly going back while the reader is preparing or as you tap a card.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/73c5162a-5b9f-4b16-b178-9671b7c12f91


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
15223 Remove option to skip location error

If we don’t have location permissions, the payment will fail. The skip option only existed to support a POS early implementation where we didn’t have designs for showing the alerts.

This removes the skip option in preparation for POS adopting the location permissions flow.

joshheald committed:
15223 Rename locationRequired action to cancel

The previous `dismiss` name didn’t make it clear that the connection or payment attempt would end, and cancel makes that clearer. I avoided `endSearch` because it’s related to payments as well as connections, potentially.

joshheald committed:
15223 Show location permission error in POS

joshheald committed:
15223 Add Location pre-alert to POS

joshheald committed:
15223 Add POS location permissions to 22.1 notes

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
15223 Make location error cancel button snappy

joshheald committed:
15223 Fix `Try again` for non-retry location error

The cancellation action from the non-retryable location error during payment never returns if the `paymentCancellable` isn’t valid.

That can happen when we call `collectPaymentMethod` with an intent which has already had a location permissions error.

By nilling the cancellable when we produce an error from the completion handler, second and subsequent retries of the location error will use the fallback approach and successfully cancel.

staskus commented on RELEASE-NOTES.txt:
```diff
@@ -7,6 +7,7 @@
 - [*] Order Creation: Fixed an issue where order recalculation would stop working after canceling a confirmation with unsaved changes [https://github.com/woocommerce/woocommerce-ios/pull/15392].
 - [internal] POS: Always create a new order during the checkout [https://github.com/woocommerce/woocommerce-ios/pull/15427].
 - [internal] Improve data fetching in Order Details, to avoid I/O performance on the main thread. [https://github.com/woocommerce/woocommerce-ios/pull/14999]
+- [internal] POS: Show location pre-alert and error during reader connection [
```
```suggestion
- [internal] POS: Show location pre-alert and error during reader connection [https://github.com/woocommerce/woocommerce-ios/pull/15456].
```

joshheald replied:
Oops! Thanks

staskus reviewed (approved):
Thank you, looks good!

The flow itself works as described and looks well. I tried it under various accessibility settings, and they all adapted well.

Thanks for also addressing the cancelation issue. I haven't been able to reproduce additional issues with various cancelation attempts.

joshheald committed:
Add missing link to PR for location flow in POS

Co-authored-by: Povilas Staskus <povilas.staskus@automattic.com>
